File: 9.palindrome-number.c
/*
 * @lc app=leetcode id=9 lang=c
 *
 * [9] Palindrome Number
 */

#include <stdbool.h>
bool isPalindrome(int x)
{
	if (x < 0 || (x != 0 && x % 10 == 0))
		return false;
	int inv = 0;
	int dig;
	int tmp = x;
	while (x != 0)
	{
		dig = x % 10;
		x /= 10;
		if (inv > 214748364 || (inv == 214748364 && dig > 7))
			return 0;
		if (inv < -214748364 || (inv == -214748364 && dig < -8))
			return 0;
		inv = inv * 10 + dig;
	}
	return inv == tmp;
}

File: 9.palindrome-number.java
/*
 * @lc app=leetcode id=9 lang=java
 *
 * [9] Palindrome Number
 *
 * https://leetcode.com/problems/palindrome-number/description/
 *
 * algorithms
 * Easy (44.99%)
 * Likes:    2058
 * Dislikes: 1500
 * Total Accepted:    848.4K
 * Total Submissions: 1.8M
 * Testcase Example:  '121'
 *
 * Determine whether an integer is a palindrome. An integer is a palindrome
 * when it reads the same backward as forward.
 * 
 * Example 1:
 * 
 * 
 * Input: 121
 * Output: true
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: -121
 * Output: false
 * Explanation: From left to right, it reads -121. From right to left, it
 * becomes 121-. Therefore it is not a palindrome.
 * 
 * 
 * Example 3:
 * 
 * 
 * Input: 10
 * Output: false
 * Explanation: Reads 01 from right to left. Therefore it is not a
 * palindrome.
 * 
 * 
 * Follow up:
 * 
 * Coud you solve it without converting the integer to a string?
 * 
 */

// @lc code=start
class Solution {
    public boolean isPalindrome(int x) {
        if (x < 0)
            return false;
        if (x < 10)
            return true;
        if (x > 2147447412)
            return false;
        int tmp = x;
        int inv = 0;
        int digit;
        while (tmp != 0) {
            digit = tmp % 10;
            tmp /= 10;
            inv = inv * 10 + digit;
        }
        return inv == x;
    }
}
// @lc code=end
